File: agenkit-scala/src/main/scala/io/agenkit/middleware/TimeoutMiddleware.scala
package io.agenkit.middleware

import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future, Promise, TimeoutException}
import scala.concurrent.duration.*
import java.util.concurrent.atomic.AtomicLong
import java.util.{Timer, TimerTask}

/** Fails the request if the inner agent does not respond within `timeout`. */
class TimeoutMiddleware(
  inner: Agent,
  timeout: FiniteDuration
) extends Agent:
  private val _processedCount = new AtomicLong(0)
  private val timer           = new Timer(true) // daemon timer

  def name: String               = s"timeout(${inner.name})"
  def capabilities: List[String] = inner.capabilities :+ "timeout"

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _processedCount.incrementAndGet()
    val promise = Promise[Message]()
    val task = new TimerTask:
      def run(): Unit =
        promise.tryFailure(new TimeoutException(s"Agent timed out after $timeout"))
        ()
    timer.schedule(task, timeout.toMillis)
    val innerFuture = inner.process(message)
    innerFuture.foreach { result =>
      task.cancel()
      promise.trySuccess(result)
      ()
    }
    innerFuture.failed.foreach { ex =>
      task.cancel()
      promise.tryFailure(ex)
      ()
    }
    promise.future

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("timeout" -> timeout.toString, "inner" -> inner.name),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/middleware/TimeoutMiddleware.java
package io.agenkit.middleware;

import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 * Middleware that enforces a per-request timeout.
 */
public final class TimeoutMiddleware implements Agent {

    private final Agent inner;
    private final Duration timeout;

    public TimeoutMiddleware(Agent inner, Duration timeout) {
        this.inner = inner;
        this.timeout = timeout;
    }

    @Override
    public String getName() { return inner.getName() + "[timeout]"; }

    @Override
    public List<String> getCapabilities() { return inner.getCapabilities(); }

    @Override
    public CompletableFuture<Message> process(Message message) {
        return inner.process(message)
                .orTimeout(timeout.toMillis(), TimeUnit.MILLISECONDS)
                .exceptionallyCompose(ex -> {
                    if (ex instanceof TimeoutException) {
                        return CompletableFuture.failedFuture(
                                new TimeoutException("agent timed out after " + timeout));
                    }
                    return CompletableFuture.failedFuture(ex);
                });
    }

    @Override
    public IntrospectionResult introspect() { return inner.introspect(); }
}
